controller/contour/job.go

```diff
@@ -0,0 +1,274 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package contour
+
+import (
+	"context"
+	"fmt"
+
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
+	oputil "github.com/projectcontour/contour-operator/util"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	jobContainerName = "contour"
+	jobImageName     = "docker.io/projectcontour/contour:main"
```

Comment:
The job should use the versioned contour container image. The reason we used main historically was that re-applying the YAML would fail on upgrades because that would attempt to modify the job spec. I don't think you gave that issue here since you are recreating the job rather than modifying it.

Reply:
@jpeach I updated the desired `Job` to use the image from the `Config` API and removed `jobImageName`. I would expect the image from `Config` is set as part of the release process. If not, then the default image tag should be updated when the operator is ready for release.